src/python/torchdistx/gossip_grad.py

```diff
@@ -0,0 +1,338 @@
+        peer_idx = node_rank ^ 2**power
+        if peer_idx >= len(state.cur_topology):
+            return -1, -1
+        return state.cur_topology[peer_idx], state.cur_topology[peer_idx]
+
+    elif state.topology == Topology.DISSEMINATION:
+        send_peer_idx = (node_rank + 2**power) % state.num_nodes
+        recv_peer_idx = (node_rank - 2**power + state.num_nodes) % state.num_nodes
+        return state.cur_topology[send_peer_idx], state.cur_topology[recv_peer_idx]
+
+
+def _gossip(state, grad, scaling_factor=0.5):
+    r"""
+    Gossiping stage.
+
+    At this step, it obtains communication peers,
+    stacks ``torch.distributed.irecv`` and ``torch.distributed.isend`` operations,
+    and performs communication with ``torch.distributed.batch_isend_irecv``.
+    Finally, received and current gradients are added together
+    and scaled appropriately, i.e. since communication happens
+    only between 2 peers at a time, summed gradients are divided
+    by 2 (or multiplied by 0.5)
+
+    For more information, please refer to the original
+    `paper <https://arxiv.org/abs/1803.05880>`_
+
+    Args:
+        state (GossipGradState): State for GossipGraD communication hook.
+        grad (torch.Tensor): A gradient for the local batch
+            that needs to be communicated across ranks.
+        scaling_facto (float): Scaling factor to apply after
+            received and current gradients are combined.
+
+    """
+    send_peer, recv_peer = _get_send_recv_peers(state)
+    if send_peer == -1:
+        return
+
+    assert send_peer is not None and recv_peer is not None, (
+        "Failed to calculate send and receive peers: "
+        f"(`send_peer` is {send_peer} and `recv_peer` is {recv_peer})"
+    )
+    # Need to check that send and receive peers are not equal to a current rank
+    assert send_peer != state.rank and recv_peer != state.rank, (
+        "Expected send and receive peers to differ from a current rank: "
+        f"(current rank is {state.rank}, `send_peer` is {send_peer}\
+        and `recv_peer` is {recv_peer})"
+    )
+    assert (
+        send_peer != -1 and recv_peer != -1
+    ), "Communication peers are not present in a current topology"
+    recv_grad = torch.empty_like(grad)
+    ops = []
+
+    # For ranks not in the `master_process_group`,
+    # `master_process_group` is an `object` instance
+    assert isinstance(
+        state.master_process_group, ProcessGroup
+    ), "`master_process_group` is not an instance of `ProcessGroup`"
+    ops.append(
+        dist.P2POp(
+            op=dist.isend, tensor=grad, peer=send_peer, group=state.master_process_group
+        )
+    )
+    ops.append(
+        dist.P2POp(
+            op=dist.irecv,
+            tensor=recv_grad,
+            peer=recv_peer,
+            group=state.master_process_group,
+        )
+    )
+    reqs = dist.batch_isend_irecv(ops)
+    for req in reqs:
+        req.wait()
+    torch.cuda.synchronize()
```

Comment:
If we need a sync here, will a stream sync (instead of device sync) be sufficient?

Reply:
@mrshenli I've consulted with Ke Wen and he confirmed that stream sync will suffice

Reply:
@mrshenli and @rohan-varma general update regarding this issue. With a patched PyTorch (a patch which fixes `batch_isend_irecv` bug I explained offline) I can get rid of this `synchronization` completely.

Reply:
Cool! I think this version should be good to land for now, and we can run experiments with the PT patch and indicate that clearly, and ensure there are clear plans to land that patch. Excellent debugging!